File: src/main/java/de/byjoker/myjfql/database/DatabaseServiceImpl.java
package de.byjoker.myjfql.database;

import de.byjoker.jfql.util.ID;
import de.byjoker.myjfql.core.MyJFQL;
import de.byjoker.myjfql.exception.FileException;
import de.byjoker.myjfql.util.FileFactory;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.IntStream;

public class DatabaseServiceImpl implements DatabaseService {

    private final FileFactory factory;
    private final List<Database> databases;

    public DatabaseServiceImpl() {
        this.factory = new FileFactory();
        this.databases = new ArrayList<>();
    }

    @Override
    public void createDatabase(Database database) {
        if (getDatabaseByName(database.getName()) != null)
            throw new FileException("Database already exists!");

        if (existsDatabase(database.getId())) {
            database.setId(ID.generateNumber().toString());
            createDatabase(database);
            return;
        }

        saveDatabase(database);
    }

    @Override
    public void saveDatabase(Database database) {
        for (int i = 0, databasesSize = databases.size(); i < databasesSize; i++) {
            if (databases.get(i).getName().equals(database.getName())) {
                databases.set(i, database);
                return;
            }
        }

        databases.add(database);
    }

    @Override
    public boolean existsDatabaseByIdentifier(String identifier) {
        if (identifier.startsWith("#"))
            return existsDatabase(identifier.replaceFirst("#", ""));

        return existsDatabaseByName(identifier);
    }

    @Override
    public boolean existsDatabaseByName(String name) {
        return databases.stream().anyMatch(database -> database.getName().equals(name));
    }

    @Override
    public boolean existsDatabase(String id) {
        return databases.stream().anyMatch(database -> database.getId().equals(id));
    }

    @Override
    public void deleteDatabase(String id) {
        databases.removeIf(database -> database.getId().equals(id));
        new File("database/" + id + ".json").delete();
    }

    @Override
    public Database getDatabaseByIdentifier(String identifier) {
        if (identifier.startsWith("#")) {
            return getDatabase(identifier.replaceFirst("#", ""));
        }

        return getDatabaseByName(identifier);
    }

    @Override
    public Database getDatabaseByName(String name) {
        return databases.stream().filter(database -> database.getName().equals(name)).findFirst().orElse(null);
    }

    @Override
    public Database getDatabase(String id) {
        return databases.stream().filter(database -> database.getId().equals(id)).findFirst().orElse(null);
    }

    @Override
    public List<Database> getDatabases() {
        return databases;
    }

    @Override
    public void loadAll() {
        loadAll(new File("database"));
    }

    @Override
    public void loadAll(File space) {
        databases.clear();

        for (File file : Objects.requireNonNull(space.listFiles())) {
            final JSONObject jsonDatabase = factory.load(file);
            final JSONArray jsonTables = jsonDatabase.getJSONArray("tables");

            final String name = file.getName().replaceFirst(".json", "");

            if (name.contains("%") || name.contains("#") || name.contains("'")) {
                MyJFQL.getInstance().getConsole().logWarning("Database used unauthorized characters in the id!");
            } else {
                final Database database = new Database(name, jsonDatabase.getString("name"));

                for (int j = 0; j < jsonTables.length(); j++) {
                    final JSONObject jsonTable = jsonTables.getJSONObject(j);

                    final Table table = new Table(jsonTable.getString("name"), null, jsonTable.getString("primary"));
                    final List<String> list = new ArrayList<>();

                    for (final Object obj : jsonTable.getJSONArray("structure")) {
                        list.add(obj.toString());
                    }

                    table.setStructure(list);

                    final JSONArray jsonColumns = jsonTable.getJSONArray("columns");

                    IntStream.range(0, jsonColumns.length()).mapToObj(jsonColumns::getJSONObject).forEach(jsonColumn -> {
                        Column column = new Column();
                        column.setCreation(jsonColumn.getLong("creation"));
                        column.setContent(jsonColumn.getJSONObject("content").toMap());
                        table.addColumn(column);
                    });

                    if (!table.getName().contains("%") && !table.getName().contains("#") && !table.getName().contains("'"))
                        database.saveTable(table);
                    else
                        MyJFQL.getInstance().getConsole().logWarning("Table used unauthorized characters in the name!");
                }

                if (!database.getName().contains("%") && !database.getName().contains("#") && !database.getName().contains("'"))
                    databases.add(database);
                else
                    MyJFQL.getInstance().getConsole().logWarning("Database used unauthorized characters in the name!");
            }
        }
    }

    @Override
    public void updateAll() {
        updateAll(new File("database"));
    }

    @Override
    public void updateAll(File space) {
        databases.forEach(database -> {
            final File file = new File(space.getPath() + "/" + database.getId() + ".json");
            final JSONObject jsonObject = new JSONObject();
            jsonObject.put("name", database.getName());
            jsonObject.put("tables", database.getTables());
            factory.save(file, jsonObject);
        });
    }

}

File: src/main/java/de/byjoker/myjfql/database/DatabaseServiceImpl.kt
package de.byjoker.myjfql.database

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ArrayNode
import com.fasterxml.jackson.databind.node.ObjectNode
import de.byjoker.myjfql.core.MyJFQL
import de.byjoker.myjfql.exception.DatabaseException
import de.byjoker.myjfql.util.Json
import kotlinx.coroutines.*
import java.io.File

@DelicateCoroutinesApi
class DatabaseServiceImpl : DatabaseService {

    private val databases: MutableMap<String, Database> = mutableMapOf()

    override fun createDatabase(database: Database) {
        if (databases.containsKey(database.id) || getDatabaseByName(database.name) != null) {
            throw DatabaseException("Database already exist!")
        }

        saveDatabase(database)
    }

    override fun saveDatabase(database: Database) {
        databases[database.id] = database
    }

    override fun deleteDatabase(id: String) {
        File("databases/$id").delete()
        databases.remove(id)
    }

    override fun getDatabase(id: String): Database? {
        return databases[id]
    }

    override fun getDatabaseByName(name: String): Database? {
        return databases.values.firstOrNull { database -> database.name == name }
    }

    override fun getDatabases(): List<Database> {
        return databases.values.toList()
    }

    override suspend fun load(backend: File): Database? {
        if (!backend.exists()) {
            return null
        }

        val console = MyJFQL.getInstance().console
        console.info("Reading ${backend.name} from hard-drive...")

        fun buildTable(tableNode: JsonNode, databaseId: String): Table {
            return when (tableNode["type"].asText()) {
                "RELATIONAL" -> RelationalTable(
                    tableNode["id"].asText(),
                    tableNode["name"].asText(),
                    databaseId,
                    tableNode["structure"].toList().map { it.asText() }.toMutableList(),
                    tableNode["primary"].asText(),
                    tableNode["partitioner"].asText(),
                )
                "DOCUMENT" -> DocumentCollection(
                    tableNode["id"].asText(),
                    tableNode["name"].asText(),
                    tableNode["partitioner"].asText(),
                    databaseId
                )
                else -> throw DatabaseException("Unknown table type!")
            }
        }

        fun buildEntry(entryNode: JsonNode, type: TableType): Entry {
            val content = entryNode["content"] as ObjectNode
            val entry: Entry = when (type) {
                TableType.DOCUMENT -> Document(entryNode["id"].asText())
                TableType.RELATIONAL -> RelationalEntry(entryNode["id"].asText())
            }
            entry.applyContent(content, fully = true)

            return entry
        }

        fun pushEntries(entriesNode: JsonNode, table: Table) {
            for (entryNode in entriesNode) {
                table.pushEntry(buildEntry(entryNode, table.type))
            }
        }

        try {
            if (backend.isFile) {
                val databaseNode = Json.read(backend)
                val database = SimpleDatabase(
                    databaseNode["id"].asText(),
                    databaseNode["name"].asText(),
                    DatabaseType.valueOf(databaseNode["type"].asText())
                )

                val tablesNode = databaseNode["tables"] as ArrayNode

                for (tableNode in tablesNode) {
                    val table = buildTable(tableNode, database.id)
                    pushEntries(tableNode["entries"] as ArrayNode, table)
                    database.pushTable(table)
                }

                return database
            }

            val databaseNode = Json.read(File("${backend.path}/%database%.json"))
            val database = SimpleDatabase(
                databaseNode["id"].asText(),
                databaseNode["name"].asText(),
                DatabaseType.valueOf(databaseNode["type"].asText())
            )

            val tables = databaseNode["tables"].toList().map { it.asText() }

            when (database.type) {
                DatabaseType.DOCUMENT -> {
                    for (tableId in tables) {
                        val tableNode = Json.read(File("${backend.path}/${tableId}.json"))
                        val table = buildTable(tableNode, database.id)
                        pushEntries(tableNode["entries"] as ArrayNode, table)

                        database.pushTable(table)
                    }
                }
                DatabaseType.SHARDED -> {
                    for (tableId in tables) {
                        val tableNode = Json.read(File("${backend.path}/${tableId}/%table%.json"))
                        val table = when (tableNode["type"].asText()) {
                            "RELATIONAL" -> RelationalTable(
                                tableNode["id"].asText(),
                                tableNode["name"].asText(),
                                database.id,
                                tableNode["structure"].toList().map { it.asText() }.toMutableList(),
                                tableNode["primary"].asText(),
                                tableNode["partitioner"].asText()
                            )
                            "DOCUMENT" -> DocumentCollection(
                                tableNode["id"].asText(),
                                tableNode["name"].asText(),
                                tableNode["partitioner"].asText(),
                                database.id
                            )
                            else -> throw DatabaseException("Unknown table type!")
                        }

                        for (entryFile in File("${backend.path}/${table.id}").listFiles()) {
                            if (entryFile.name == "%table%.json") continue
                            table.pushEntry(buildEntry(Json.read(entryFile), table.type))
                        }

                        database.pushTable(table)
                    }
                }
                else -> return null
            }

            console.info("Finished ${backend.name} ✓")
            return database
        } catch (ex: Exception) {
            console.error("Failed ${backend.name} ×")
        }

        return null
    }

    override fun loadAll() {
        GlobalScope.async {
            withContext(Dispatchers.Default) {
                File("databases").listFiles().forEach { file -> load(file)?.let { createDatabase(it) } }
            }
        }
    }

    override suspend fun write(backed: File, t: Database) {
        val console = MyJFQL.getInstance().console
        console.info("Writing ${t.name} to hard-drive...")

        try {
            backed.mkdirs()

            when (t.type) {
                DatabaseType.STANDALONE, DatabaseType.INTERNAL -> {
                    Json.write(t, File("${backed.path}/${t.id}.json"))
                }
                DatabaseType.DOCUMENT -> {
                    File("${backed.path}/${t.id}").mkdirs()
                    Json.write(DatabaseRepresentation(t), File("${backed.path}/${t.id}/%database%.json"))

                    t.getTables().forEach { table ->
                        Json.write(table, File("${backed.path}/${t.id}/${table.id}.json"))
                    }
                }
                DatabaseType.SHARDED -> {
                    File("${backed.path}/${t.id}").mkdirs()
                    Json.write(DatabaseRepresentation(t), File("${backed.path}/${t.id}/%database%.json"))

                    t.getTables().forEach {
                        File("${backed.path}/${t.id}/${it.id}").mkdirs()
                        Json.write(
                            TableRepresentation(it),
                            File("${backed.path}/${t.id}/${it.id}/%table%.json")
                        )

                        it.getEntries().forEach { entry ->
                            Json.write(entry, File("${backed.path}/${t.id}/${it.id}/${entry.id}.json"))
                        }
                    }
                }
            }
        } catch (ex: Exception) {
            console.error("Failed ${t.name} ×")
            return
        }

        console.info("Finished ${t.name} ✓")
    }

    override fun writeAll() {
        GlobalScope.async {
            withContext(Dispatchers.Default) {
                databases.values.forEach { write(File("databases"), it) }
            }
        }
    }

}
